mediatomb: Improve service + add gerbera support and tests

ardumont opened the pull request:
# Motivations

- Allow service to run with the most recent maintained mediatomb fork (gerbera)
- Add more configuration options (e.g. media directories to autoscan, ...)
- Leverage options to make the configuration lazy
- Fix `transcoding` option which if enabled installs the necessary dependencies to actually work
- Add tests to the service (there was none), testing scenario are:
  1. service runs with existing mediatomb package
  2. service runs with the new gerbera package
- Add types to existing service options which did not declare one
- Open a new option `openFirewall` (defaults to false) and improve the firewall declaration using it.
- Add release note (20.09) information about new options and breaking change (`openFirewall` off by default now)

# Actions

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS (`Linux alderaan 5.7.7 #1-NixOS SMP Tue Jun 30 20:21:22 UTC 2020 x86_64 GNU/Linux`)
   - [ ] macOS
   - [x] other Linux distributions: debian with nix (`Linux yavin4 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07) x86_64 GNU/Linux`)
- [x] Added tests as none existed
- [x] Tested compilation of all pkgs that depend on this: nixpkgs-review pr 93450
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

# Notes

I built the diff on the started works [1] [2]

[1] #82429 (apparently idle, asked but got no reply so i continued from it)

[2] #93226 (my own work about gerbera, merged now)

[4]

```
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/93450/head:refs/nixpkgs-review/1
$ git worktree add /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs 9fab6f9b2b2d8927d3984a7549ed4c56829d2065
Preparing worktree (detached HEAD 9fab6f9b2b2)
Updating files: 100% (22851/22851), done.
HEAD is now at 9fab6f9b2b2 Merge pull request #98823 from r-ryantm/auto-update/multimon-ng
$ nix-env -f /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit 2c4cf27d61b07dfbac3ffa73658471c8180eb972
Automatic merge went well; stopped before committing as requested
$ nix-env -f /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs -qaP --xml --out-path --show-trace --meta
Nothing to be built.
https://github.com/NixOS/nixpkgs/pull/93450
$ nix-shell /home/tony/.cache/nixpkgs-review/pr-93450-4/shell.nixnixpkgs-review pr 93450
$ git -c fetch.prune=false fetch --force https://github.com/NixOS/nixpkgs master:refs/nixpkgs-review/0 pull/93450/head:refs/nixpkgs-review/1
$ git worktree add /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs 9fab6f9b2b2d8927d3984a7549ed4c56829d2065
Preparing worktree (detached HEAD 9fab6f9b2b2)
Updating files: 100% (22851/22851), done.
HEAD is now at 9fab6f9b2b2 Merge pull request #98823 from r-ryantm/auto-update/multimon-ng
$ nix-env -f /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs -qaP --xml --out-path --show-trace
$ git merge --no-commit 2c4cf27d61b07dfbac3ffa73658471c8180eb972
Automatic merge went well; stopped before committing as requested
$ nix-env -f /home/tony/.cache/nixpkgs-review/pr-93450-4/nixpkgs -qaP --xml --out-path --show-trace --meta
Nothing to be built.
https://github.com/NixOS/nixpkgs/pull/93450
$ nix-shell /home/tony/.cache/nixpkgs-review/pr-93450-4/shell.nix
```

Cheers,

edwtjo committed:
mediatomb: make service compatible with the gerbera fork

The duplication of the interface xml tag is needed for
the daemon to respect the setting.

ardumont committed:
mediatomb/gerbera: Make transcoding option lazy and runnable if activated

In the sense that the pkgs dependency will be pulled if the service is
transcoding enabled. Otherwise, the transcoding part is completely dropped from
the generated configuration.

ardumont committed:
mediatomb/gerbera: Introduce the pcDirectoryHide option

ardumont committed:
mediatomb/gerbera: Introduce the mediaDirectories option

So users can declare their autoscan directories configuration from nix.

ardumont committed:
mediatomb/gerbera: Add missing types to options

This also fixes some various small limitations:
- Drop unnecessary quoting
- Drop duplicated gerbera interface definition
- Fix configuration indentation

ardumont committed:
mediatomb/gerbera: Make the actual configuration generation lazy

Also use verbose flag in cli command to make the intent clearer.

ardumont committed:
mediatomb/gerbera: Bootstrap tests on service

This exposes 2 scenario running the mediatomb service:
- one running with the unmaintained mediatomb package
- one running with the new maintained gerbera package

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -256,31 +325,33 @@ in {
-    users.groups = optionalAttrs (cfg.group == "mediatomb") {
-      mediatomb.gid = gid;
+    users.groups = {
+      "${cfg.group}".gid = gid;
```
The group should probably become an enum or something instead here.
But I don't know yet how to do that ¯\_(ツ)_/¯

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -158,21 +196,30 @@ in {
       ps3Support = mkOption {
         type = types.bool;
-        default = false;
+        default = true;
```
oops, I missed that I flipped that. I'll fix.

ardumont commented:
I just:
- fixed a flipped flag option to keep its existing value
- clarified a docstring

aanderse commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -6,37 +6,97 @@ let
-  mtConf = pkgs.writeText "config.xml" ''
-  <?xml version="1.0" encoding="UTF-8"?>
-  <config version="2" xmlns="http://mediatomb.cc/config/2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://mediatomb.cc/config/2 http://mediatomb.cc/config/2.xsd">
+  transcodingConfig = if cfg.transcoding then with pkgs; ''
```
It has been too many years since I have used `mediatomb` to recall all the details, but would it be better for us to do xml substitutions on the configuration file provided by upstream instead of generating our own? I'm not sure what the right answer is here but would like to hear your thoughts.

ardumont replied:
Well, as far as I could tell, they don't provide one. The new gerbera cli
provide the means to generate one though but that hardcodes values... (thus not
working for nixos's case).

So I think continuing that way is fine. The benefit from doing this ourselves
allows to be consistent between what the user declares and the generated
configuration.

We can also benefit from nix's laziness (like I did when fixing the transcoding
option, if we don't activate it here, the default, we don't actually pull vlc
or ffmpeg ;)

The only drawback I see continuing that way is well maintenance here ¯\_(ツ)_/¯ (which should not that much of a problem if we continue maintaining this with tests (maybe ;)

ardumont replied:
Note that, as another example, I also leveraged nix's laziness here to stop completely the configuration file generation if the user enables customCfg (which as far as I understood it was not the case before ;).

aanderse reviewed (dismissed):
Here are some initial comments. I think I need more time to do a thorough review. Will circle back to this in the next few days hopefully.

ardumont commented on nixos/modules/misc/ids.nix:
```diff
@@ -228,7 +228,7 @@ in
       grafana = 196;
       skydns = 197;
-      # ripple-rest = 198; # unused, removed 2017-08-12
+      gerbera = 198;
```
Thanks for the pointer ;)

ardumont commented:
> Here are some initial comments. I think I need more time to do a thorough review. Will circle back to this in the next few days hopefully.

Awesome, thanks.
Will have a look and try to my best to answer the concerns and fix accordingly those ;)

Cheers,

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -256,31 +326,33 @@ in {
-      path = [ pkgs.mediatomb ];
-      serviceConfig.ExecStart = "${pkgs.mediatomb}/bin/mediatomb -p ${toString cfg.port} ${if cfg.interface!="" then "-e ${cfg.interface}" else ""} ${if cfg.customCfg then "" else "-c ${mtConf}"} -m ${cfg.dataDir}";
+      serviceConfig.ExecStart = "${binaryCommand} --port ${toString cfg.port} ${interfaceFlag} ${configFlag} --home ${cfg.dataDir}";
       serviceConfig.User = "${cfg.user}";
```
Indeed, thanks.

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -256,31 +326,33 @@ in {
       serviceConfig.User = "${cfg.user}";
     };
 
-    users.groups = optionalAttrs (cfg.group == "mediatomb") {
```
Thanks for the heads up, sounds reasonable an approach.

I'm still trying to wrap my head around the other parts of the equation:
- about dynamic/static uid/gid you referred to me...
- also migration, how one can run this without keeping too much `mediatomb` references all around?... etc... [1]

[1] All in all, it looks like simplest way forward is to keep mediatomb user/group even when running with the new gerbera package...

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,20 +253,28 @@ in {
         '';
       };
 
       user = mkOption {
```
I only added types to the new options I introduced.
But sure, I can update the ones which I tried to improve their docstring upon (no idea if I succeeded though ;).

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -256,31 +326,33 @@ in {
     };
 
-    networking.firewall = {
+    networking.firewall.interfaces."${cfg.interface}" = {
```
I'm a bit uneasy with that one by the way.
It's coming from the commit I started upon.

I don't really know how to deal with the case where the `cfg.interface` is not specified (which is the default case, as empty string...).

Do you have any hints here?

Thanks in advance.

ardumont replied:
Also I read we should not deal with the firewall rules within the service.
I kept this though as it's already the existing behavior...
(And the PR is getting a "tad" big already... ;)

aanderse replied:
Do a quick `grep` for `openFirewall`. For this service I would suggest adding an `openFirewall` option and then do something like this (maybe a bit cleaner, though :smile:):
```
networking.firewall = mkMerge [
  (mkIf (cfg.openFirewall && cfg.interface != "") {
    interfaces."${cfg.interface}" = {
      allowedUDPPorts = [ 1900 cfg.port ];
      allowedTCPPorts = [ cfg.port ];
    };
  })
  (mkIf (cfg.openFirewall && cfg.interface == "") {
      allowedUDPPorts = [ 1900 cfg.port ];
      allowedTCPPorts = [ cfg.port ];
  })
];
```

ardumont replied:
Oh yeah, thanks!
That's clearer (pushed a new commit with this, I removed the set duplication but that's about it)

ardumont commented:
(NdA: PR description editing)
> [x] Added tests as none existed (note that they pass when run interactively thoughnix-build mediatomb.nix -A driver && ./result/bin/nixos-test-driver)

I can run interactive tests when proper rights are set to `/dev/kvm` (most probably issue with nix within debian... i'm using kvm from debian and not the nix one because i don't recall being able to do otherwise...)
Anyway, amending the description now ;)

ardumont committed:
mediatomb/gerbera: Improve firewall rules and open firewall option

This changes the default behavior which opened by default the firewall rules.
The users now need to declare explicitely they want to open the firewall.

ardumont commented:
> All checks have failed

lol, 1 check failed.
I missed a `toString` call within a docstring...

>  ardumont force-pushed the ardumont:gerbera-service  branch from 7e674dd to 762af6d 2 minutes ago

This should be fixed now.

aanderse commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,83 +257,129 @@ in {
 
+      openFirewall = mkOption {
+        type = types.bool;
+        default = true;  # On by default for retro-compatibility with existing behavior
```
Hmm... we may want to check on this. I can't recall the details. I think we're supposed to have all ports closed unless explicitly allowed by the sysadmin. **If** so an entry in the release notes would suffice here.

ardumont replied:
Yes, i recall a discussion (github issue) about services should stop opening firewall themselves... (let me search for it... found a rule in a doc instead [1])

I specifically left it open here because it was the existing behavior i did not want to break (thus the comment so reviewer know my intent). If "breaking the existing behavior" is acceptable with an update in the release note though (i don't really know what that is yet ;) then i'll adapt accordingly.

[1] https://nixos.org/nixpkgs/manual/#reviewing-contributions-new-modules (it's for a new module though ¯\_(ツ)_/¯)

Thanks again

ardumont replied:
So, the option is false by default now. And I added a note in the release-note (20.09) in the breaking-compatibility section about it.

aanderse reviewed (commented):
Sorry I'm going to need some more time to review this. I'm hoping someone else can jump in and offer some review as well because it has been too many years since I've used `mediatomb` to be the only person reviewing this.

ardumont commented:
Hello,

> Sorry I'm going to need some more time to review this.

Well, sure ;)

For information, I separated logically the commits so the review is easier one
by one. I don't really know how to simplify in another way.

> I'm hoping someone else can jump in and offer some review as well because it
> has been too many years since I've used mediatomb to be the only person
> reviewing this.

Hopefully, yes. That'd be great, the more eyes to look, the more improvments we can do ;)

In the mean time, do you have some time to review the less impactful diffs (the
ones the diff is built upon here)?

- #93226 (gerbera init at 1.5.0)
- #93048 (libupnp 1.12 -> 1.12.1)

----

> ... to be the one reviewing this...

Yes, that's why I opened tests around this. After touching so much part on
this, I wanted to prove I did not break anything both to the reviewers and me.
(Personally, I'm confident now ;)

So in those tests, I'm making sure both old and new mediatomb service still
starts.

Note that I did not add all the possible scenarios though. I was a bit wary of
the time too many checks would take.

I could still improve the testing, so we actually check:

- some part of the generated configuration (the parts that makes sense, for
  example the mediaDirectories, transcoding, ...).

- one "pattern" we can grep for in the service logs "Configuration check
  succeeded." This ones ensures the service will start (when it fails, the
  service is in failed state).

- "The Web UI can be reached by following this link:
  http://${interface}:${toString port}/" as well [1]. This ones actually is
  dependent on the declaration about interface and port so i'd say it's
  relevant.

[1] web ui is hard-coded to on by default. I did not open an option for it
because, well, the "diff" is long already...

Cheers,

ardumont commented:
> @ardumont mediatomb: Improve firewall rules + open firewall option (off by defa…  … 6633914

Note for self, look into adding a release note [1] to explicit the firewall rule option change [2]

[1] nixpkgs/nixos/doc/manual/release-notes/rl-2009.xml <- in that file probably

[2] as of this commit, the service no longer opens the firewall rules by default, this needs an explicit user declaration `openFirewall = true;`

ardumont committed:
mediatomb/gerbera: Add release note information for 20.09

Note that it made into 2 entries, one about new options in the first section.
Another in the breaking compatibility section due to the openFirewall option
which changes the behavior.

ardumont commented:
Do I need to rebase and fix conflicts now?

aanderse commented:
@ardumont sure. We really should find someone who uses `mediatomb` to test and review this...

ardumont commented:
> @ardumont sure. 

done.
everything is still fine (build and tests).

> We really should find someone who uses mediatomb to test and review this...

It'd be neat because since the release notes got updated, I gather conflicts will happen more often...

I note you are still noted as requesting changes. But as far as I know and remember, I addressed all your points, didn't I?

;)

aanderse reviewed (commented):
Have we found anyone to review this yet? Did you post on https://discourse.nixos.org/?

timokau commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -195,7 +195,30 @@ GRANT ALL PRIVILEGES ON *.* TO 'mysql'@'localhost' WITH GRANT OPTION;
+   <listitem>
+     <para>
+       The <link linked="opt-services.mediatomb">mediatomb service</link>
+       declares new and adapt some options so the configuration generation is
```
```suggestion
       declares some new and adapts some of the old options so the configuration generation is
```

Grammar seemed off, but I might be wrong (I'm not a native speaker myself).

ardumont replied:
Thanks, it's off alright.
Improving.

timokau commented on nixos/doc/manual/release-notes/rl-2009.xml:
```diff
@@ -743,6 +766,23 @@ CREATE ROLE postgres LOGIN SUPERUSER;
+       </programlisting>
+       One new option <literal>openFirewall<literal> has been introduced which
+       defaults to false. If you relied on the service to open the firewall
+       rules itself, you should now declare it with:
```
```suggestion
       itself, you should now declare it with:
```

(I think it should either be "open the firewall" or "add the firewall rules", but not a combination of both).

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -6,37 +6,97 @@ let
+      path = mkOption {
+        type = types.str;
+        description = ''
+          Absolute directory path to the media directory to index
```
All the descriptions should add with a period. Unfortunately I didn't find any formal docs to reference here, but it seems to be the de-facto standard (looking at `man configuration.nix`).

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -108,46 +171,25 @@ let
-  '';
+'';
+  defaultFirewallRules = {
+    allowedUDPPorts = [ 1900 cfg.port ];
```
Why 1900?

ardumont replied:
beats me, I'm not the one designing mediatomb/gerbera...
Also, note that I'm keeping the existing behavior... ¯\_(ツ)_/¯

timokau replied:
Alright. Just seemed odd that there is one configurable port and then another hardcoded one.

ardumont replied:
yes, indeed. 

I triple-checked. It seems there is nothing we can do about it (according to my understanding of the doc [1]).

Quoting:
```
You should also make sure that your firewall is not blocking port UDP port 1900 (required for SSDP) and UDP/TCP port of Gerbera. By default Gerbera will select a free port starting with 49152, however you can specify a port of your choice in the configuration file.
```

[1] http://docs.gerbera.io/en/latest/run.html?highlight=udp%20port#network-setup

timokau reviewed (changes requested):
Some nitpicks.

Since no qualified reviewer is showing up, I will review this and merge it once it looks good to me. I am not a mediatomb user, but it seems like you have done more than enough testing yourself. If anything breaks, people can complain later. There's plenty of time for that until the next release.

ardumont commented:
Thanks for the review!

I will rebase, fix conflicts and check for adaptations a bit later in the day.

Cheers,

ardumont commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,83 +257,129 @@ in {
     };
 
-    networking.firewall = {
-      allowedUDPPorts = [ 1900 cfg.port ];
```
@timokau existing behavior ^ ;)

ardumont commented:
> I will rebase, fix conflicts and check for adaptations a bit later in the day.

done.

I have:
- fixed the conflicts
- adapted according to your remarks (well aside the 1900 port which i can't do anything about). And marked as resolved the related discussions.
- run build and unit tests and everything is still fine.

Cheers,

ardumont commented:
I rebased on other branches to make the update the base source code.
I took that opportunity to fix again a conflict which happened in between updates...

ardumont commented:
Rebased to latest master which now includes the base commits this needed to build fine.

@timokau Is is a reasonable time enough for me to squash some commits now?

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -6,37 +6,97 @@ let
-  <?xml version="1.0" encoding="UTF-8"?>
-  <config version="2" xmlns="http://mediatomb.cc/config/2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://mediatomb.cc/config/2 http://mediatomb.cc/config/2.xsd">
+  transcodingConfig = if cfg.transcoding then with pkgs; ''
+ <transcoding enabled="yes">
```
The whole string-block should be indented one more level.

ardumont replied:
yeah, my bad. I should have looked again at my existing code with my new knowledge prior to pinging you, sorry about that.

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -108,46 +171,25 @@ let
-        </profile>
-      </profiles>
-    </transcoding>
+${transcodingConfig}
```
Indentation (keep in mind it is ["intelligently stripped"](https://nixos.org/manual/nix/stable/#idm140737322059168) from multiline strings).

ardumont replied:
Thanks for the link.

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,83 +257,129 @@ in {
+        default = "/var/lib/${name}";
         description = ''
-          The directory where mediatomb stores its state, data, etc.
+          The directory where ${name} stores its state, data, etc.
```
I'd call this "packageName" or something. With just "name" I would expect it to refer to the service, something like "mediatomb-service".

ardumont replied:
I won't rename name because i don't want to change too much the existing part.
I settled on using {serverName} in the description, does that sound good enough?

timokau commented on nixos/tests/mediatomb.nix:
```diff
@@ -0,0 +1,81 @@
+    port = 49152
+    serverGerbera.succeed("mkdir -p /var/lib/mediatomb/{pictures,audio}")
+    serverGerbera.succeed("chown -R mediatomb:mediatomb /var/lib/mediatomb")
+    # service running gerbera fails the first time claiming something is already bound
```
This is probably since both services want to bind to udp 1900, which is not configurable.

ardumont replied:
oh yeah, thanks for pointing it out, makes total sense now.

ardumont replied:
mmm, no, my understanding was that it was 2 different instances...

Anyway, I tried to remove that workaround and it seems to work without it... ¯\_(ツ)_/¯
So i'll clean up.

ardumont replied:
> Anyway, I tried to remove that workaround and it seems to work without it... ¯_(ツ)_/¯

No (i'm tired it seems)... I only removed the `sleep 2` instruction, not the workaround (`systemctl restart mediatomb`)  and it still works without it.
Removing the workaround and then, back to square one with failure

ardumont replied:
I kept it as is in the end.

timokau reviewed (changes requested):
Gave this a more thorough review now. Yes, I think when this is resolved it would be time to clean up the history.

ardumont commented:
> Gave this a more thorough review now. Yes, I think when this is resolved it would be time to clean up the history.

Thanks, i had a shot at this.
I have:
- rebased and squashed commits into logical blocks
- reworded a bit most commit messages
- fixes again some typos here and there
- adapted according to your reviews what I could

Hoping this is enough ;)

Thanks in advance.

Cheers,

timokau reviewed (changes requested):
ofBorg eval fails:

> undefined variable 'serverName' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/packet-spot-eval-2/nixos/modules/services/misc/mediatomb.nix:264:33

ardumont commented:
> ofBorg eval fails:
> undefined variable 'serverName' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/packet-spot-eval-2/nixos/modules/services/misc/mediatomb.nix:264:33

Thanks and sorry, I was sleeping...

I don't understand why I don't reproduce this... When running either the build
or the tests locally (because I did run them ;) [1]

It's not clear to me why `{name}` evals but `{serverName}` does not. I have
added a `{cfg.serverName}` and pushed as a tryout.


[1] Then again, I don't understand either why I have the following error locally
(fairly recent) and this does not show up in the output build in the first
place:

```
error: Alias spidermonkey is still in all-packages.nix
(use '--show-trace' to show detailed location information)
```

To work around this ^ (which i know is not so great ¯\_(ツ)_/¯), I have locally
removed `spidermonkey` from the `buildInputs` from the `mediatomb` derivation
itself (and then build and tests are no longer blocked).

ardumont commented:
> It's not clear to me why {name} evals but {serverName} does not. I have
> added a {cfg.serverName} and pushed as a tryout.

It works ;) 

I'm still unclear on the other part though [1]

timokau commented:
> > ofBorg eval fails:
> > undefined variable 'serverName' at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/packet-spot-eval-2/nixos/modules/services/misc/mediatomb.nix:264:33
> 
> Thanks and sorry, I was sleeping...

Absolutely no need to apologize :) The iteration speed here is already breakneck for open source standards ;)

> I don't understand why I don't reproduce this... When running either the build
> or the tests locally (because I did run them ;) [1]

Yeah this is a bit confusing at first: Its not about the build, but the evaluation. ofBorg tries to evaluate all of nixpkgs, while the simple build only lazily evaluates whatever it needs.

To avoid these issues, I have this hook in my `nixpkgs/.git/hooks/pre-push`:

```
#!/bin/sh
echo "Checking evaluation";
exec nix-env -f . -qaP --meta --json --drv-path --show-trace >/dev/null
```

I think its not quite the same ofBorg does, but it usually catches evaluation errors.

> It's not clear to me why `{name}` evals but `{serverName}` does not. I have
> added a `{cfg.serverName}` and pushed as a tryout.

Looks like that worked.

> [1] Then again, I don't understand either why I have the following error locally
> (fairly recent) and this does not show up in the output build in the first
> place:
> 
> ```
> error: Alias spidermonkey is still in all-packages.nix
> (use '--show-trace' to show detailed location information)
> ```
> 
> To work around this ^ (which i know is not so great ¯_(ツ)_/¯), I have locally
> removed `spidermonkey` from the `buildInputs` from the `mediatomb` derivation
> itself (and then build and tests are no longer blocked).

This is because `spidermonkey` is defined in both `pkgs/top-level/aliases.nix` (as an alias to `spidermonkey_68`) and `pkgs/top-level/all-packages.nix` (as an alias to `spidermonkey_38`). I'm not sure what is going on there. As far as I'm aware, `aliases.nix` is for name aliases `(black = python3.pkgs.black)` while version aliases `spidermonkey = spidermonkey_68` are usually kept in all-packages. 

That knowledge might be outdated, but either way one of the aliases should be removed.

This doesn't show up in ofBorg because this actually *is* a build error, not an eval error (and it didn't try to build `mediatomb` here).

timokau commented:
`spidermonkey` was added to `aliases.nix` in 946369adbd53, as an afterthought while deprecating `spidermonkey_38`. As a fix, I suggest you remove `spidermonkey` from `aliases.nix` and instead update the `all-packages.nix` version to `spidermonkey_68`. Its not really your bug and unrelated to this PR, so you can also leave it be if you want. But why not fix it while we're at it.

(CC @mkg20001)

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -6,37 +6,97 @@ let
+'';
+
+  configText = optionalString (! cfg.customCfg) ''
+<?xml version="1.0" encoding="UTF-8"?>
```
Indentation is still a bit off here. But I seen no other issues and its not that important.

timokau commented:
Thank you for your patience and your diligence here :rocket:

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,64 +259,107 @@ in {
-        default = "/var/lib/mediatomb";
+        default = "/var/lib/${name}";
+        description = ''
+          The directory where ${cfg.serverName} stores its state, data, etc.
```
Unfortunately I just noticed this post-merge. How can you refer to the value of an option within an option description? Have you tried building the docs with this? I suspect this will just always expand to the default value.

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,64 +259,107 @@ in {
+        default = false;
+        description = ''
+          If false (the default), this is up to the user to declare the firewall rules.
+          If true, this opens the 1900 (tcp and udp) and ${toString cfg.port} (tcp) ports.
```
Same here: Reference to an option value within an option description.

timokau commented on nixos/modules/services/misc/mediatomb.nix:
```diff
@@ -206,64 +259,107 @@ in {
         default = false;
         description = ''
-          Allow mediatomb to create and use its own config file inside ${cfg.dataDir}.
+          Allow ${name} to create and use its own config file inside ${cfg.dataDir}.
```
Also here. You should probably refer to the options by name instead.

timokau commented:
It seems like this breaks the manual build. I'm working on a fix.

timokau commented:
https://github.com/NixOS/nixpkgs/pull/100034

ardumont commented:
> Thank you for your patience and your diligence here

Well, thanks for your input on this ;)

> How can you refer to the value of an option within an option description?

I actually saw this in other derivations (I think).

> Have you tried building the docs with this?

I thought this was covered by the actual build process.
So I probably did not then, sorry ¯\_(ツ)_/¯.

How does one build the documentation? Is that [1]?

[1] https://nixos.org/manual/nixpkgs/stable/#chap-contributing

> Reference to an option value within an option description.

Ack, I'll stop doing that then.

Cheers,

mkg20001 commented:
@timokau What do you mean by updating spidermonkey to spidermonkey_68 in all-packages.nix? Should I move the alias there? Or simply drop spidermonkey alias for good?

ardumont commented:
@mkg20001 Hello, i'm pretty sure @timokau meant #100026 since that PR landed ;)

Cheers,

mkg20001 commented:
ah ok, thx

timokau commented:
> > Have you tried building the docs with this?
> 
> I thought this was covered by the actual build process.
> So I probably did not then, sorry ¯_(ツ)_/¯.

No worries. If anything, its a failure of automated testing and a weird doc format (hopefully not much longer, https://github.com/NixOS/nixpkgs/pull/88488 and https://github.com/NixOS/rfcs/pull/72).

> How does one build the documentation? Is that [1]?
> 
> [1] https://nixos.org/manual/nixpkgs/stable/#chap-contributing

I built the manpages like this: `cd nixos; nix-build release.nix --attr manpage`. But the way its documented probably also works.

worldofpeace commented:
Hi, so this touched the `20.09` release notes on `master` but I don't think this change is going to be backported. I'm going to revert the release note entirely for now as it causes conflicts on `release-20.09`.

timokau commented:
Sorry for missing that in the review, I didn't check the filename. @ardumont if you want to re-add it, you can add it to the 21.03 release notes since it will be included in that release.

ardumont commented:
> Sorry for missing that in the review, I didn't check the filename. 

Well, i did not realize it either i was targetting the wrong release.

> @ardumont if you want to re-add it, you can add it to the 21.03 release notes since it will be included in that release.

ok, i'll have a look later in the day.
Thanks.

ardumont commented:
/marvin opt-in
